che-incubator/quarkus-api-example
# Add keycloak authentication for food POST endpoint

Signed-off-by: David Kwon <dakwon@redhat.com>

There is caveat to this PR for now. You need to run:
```
oc apply -f https://raw.githubusercontent.com/dkwon17/quarkus-api-example/keycloak-devfile/template/keycloak/configMap.yaml
```
before starting the workspace. This is because this configMap contains the required Keycloak realm details.
If this configMap is not provided, the Keycloak authentication flow will not work.

This PR adds keycloak authentication for the `/food` `POST` endpoint.

### How to test this PR:
1. Run the following to apply the quakrus realm config map to mount to all workspace containers:
```
oc apply -f https://raw.githubusercontent.com/dkwon17/quarkus-api-example/keycloak-devfile/template/keycloak/configMap.yaml
```

2. Start a workspace with this url:
```
{CHE-HOST}/#https://github.com/dkwon17/quarkus-api-example/tree/keycloak-devfile
```

3. Run the `startdev` command to start the Quarkus application
4. Go to the swagger ui by accessing: `{ROUTE-HOST}/q/swagger-ui/`. (Make sure you are using `http` and not `https` here)
5. Try to test the `/food` `POST` endpoint and confirm that you get a 401 error:
<img width="1395" alt="image" src="https://user-images.githubusercontent.com/83611742/186212410-dbba7815-a100-462e-a004-95fbe38aa7be.png">
6. Press the `Authorize` button and set the `client_id` to `backend-service`. For the Keycloak login page, set the username and password to `admin`. After signing in, you should see:
<img width="1395" alt="image" src="https://user-images.githubusercontent.com/83611742/186212795-1d86cef4-3e3f-46e3-a6ad-05136a3329e4.png">

7. Try the `/food` `POST` endpoint again (remember to remove the `id` in the request body)
8. Should get 201 response code:
<img width="1379" alt="image" src="https://user-images.githubusercontent.com/83611742/186213162-581aa273-3019-46fe-b1e5-a05a5e10e1c8.png">
